README.md
# HyperGard

[![Build Status](https://travis-ci.org/Comcast/hypergard.svg?branch=master)](https://travis-ci.org/Comcast/hypergard)

Javascript client for HAL APIs, with support for Hypermedia Forms


## Installation

When using [npm](https://www.npmjs.com/)

```bash
npm install --save hypergard
```

## Usage

### Initialize HyperGard instance
```javascript
const HOMEPAGE_ENDPOINT = 'https://hypermedia-endpoint.com/'
const HalApi = new HyperGard(HOMEPAGE_ENDPOINT, {});
```

### Fetch Homepage
```javascript
const homepageResource = await HalApi.fetchHomepage();
```

### Options

#### `cacheHomepage`

Default value: `false`

Determines whether to keep locally closed over reference to homepage response, since the Homepage resource should be highly cache-able by Hypermedia standards.

#### `preloadHomepage`

Default: `true`

Auto-fetch homepage endpoint on initialization of HyperGard object.

#### `xhr`

Network request that will be passed along to [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

Additionally `hypergard` implements a timeout around any network fetch, which will default to `60000` milliseconds

```js
{
  // Headers to be added to each network fetch
  headers: {
    ['X-Custom-Header']: "value",
  }
  // Timeout period in milliseconds
  timeout: 2000,
}
```

#### `applyMiddlewareStack`

Allows you to pass an array of middleware function to wrap around each fetch.

Each middleware function should:
* Have a method signature with arguments `url`, `options`, and `next`
* Return the remaining stack `return next(url, options)` so the promise chain isn't broken

#### Example of Middleware for accessing calls before fetch. (Replaces `beforeFetch` and `uniqueFetchHeaders` functionality)
```
function setCustomHeader(url, options, next) {
  var newOptions = Object.assign({}, options, {
    headers: {
      'X-MoneyTrace': 'hey nowwwww',
    }
  });

  // Call next piece of middleware
  return next(url, newOptions);
}
```

#### Example of Middleware for accessing calls after fetch  (Replaces `afterFetchSuccess` and `afterFetchFailure` functionality)
```
function loggerMiddleware(url, options, next) {
  // Call next piece of middleware
  var promiseChain = next(url, options);

  // Log any 401
  promiseChain.catch(function(error) {
      if (error.status === '401') {
        mockLogger.log('Unauthorized', {error: error});
      }
    })

  // Return un-caught promise chain
  return promiseChain;
}
```

#### Example of Applying middleware stack

Middleware can be applied to an initialized `HyperGard` object, and will be executed based on order of array.

```
HalApi.applyMiddlewareStack([
  setCustomHeader,
  loggerMiddleware,
]);
```

## Running Tests

### Install Dependencies

```
$ npm run setup
```

### Running tests to manually validate a patchset

```
$ npm test
```

### Running tests during development

```
$ gulp test
```

That `gulp test` command will load up Chrome. Click the "Debug" button and then open the JavaScript Console to see the test results. You can also use `console` methods to be able to debug your tests.

Note that if you make a code change, you cannot simply reload http://localhost:8080/debug.html in Chrome. You have to stop the `gulp test` process with `Control+C` and then rerun the command (there's probably a better way to handle that, but it does the job for now).


CHANGELOG.md
# Changelog

This should be updated on every major tag.

Add your new version to the top and a brief summary.

**This does not replace `tag` and `release` and semver best practices. It is a supplement.**

---

## v6.0.0

Simplify build process, by no longer requiring developer to check in `dist` folder.
Upgrade node from 6 > 10. (Node 12 would require additional changes to the build process)

---

## v5.1.0

Initial publishing of open source version of `Hypergard`

---

test/setup.js
if (!Function.prototype.bind) {
  Function.prototype.bind = function(oThis) {
    if (typeof this !== 'function') {
      // closest thing possible to the ECMAScript 5
      // internal IsCallable function
      throw new TypeError('Function.prototype.bind - what is trying to be bound is not callable');
    }

    var aArgs = [].slice.call(arguments, 1),
      fToBind = this,
      fNOP = function() {
      },
      fBound = function() {
        return fToBind.apply(this instanceof fNOP
            ? this
            : oThis,
          aArgs.concat([].slice.call(arguments)));
      };

    fNOP.prototype = this.prototype;
    fBound.prototype = new fNOP();

    return fBound;
  };
}

var toString = {}.toString;

var testEndpoint = '/endpoint/';
var testOptions = {
  preloadHomepage: false,
  cacheHomepage: false,
  debug: true,

  xhr: {
    headers: {
      Authorization: 'auth'
    }
  }
};

beforeEach(function() {
  this.hyperGard = null;
  this.error = null;
  this.data = null;
  this.actionData = null;
  this.actionError = null;

  this.homepageData = function(responseData, status) {
    return new window.Response(JSON.stringify(responseData), {
      status: status || 200,
      headers: {
        'Content-type': 'application/hal+json'
      }
    });
  };

  this.onError = function(error) {
    this.error = error;
  }.bind(this);

  this.onSuccess = function(response) {
    this.data = response.data;
  }.bind(this);

  this.onActionError = function(error) {
    this.actionError = error;
  }.bind(this);

  this.onActionSuccess = function(response) {
    this.actionData = response.data;
  }.bind(this);

  spyOn(window, 'fetch');
  spyOn(this, 'onError').and.callThrough();
  spyOn(this, 'onSuccess').and.callThrough();
  spyOn(this, 'onActionError').and.callThrough();
  spyOn(this, 'onActionSuccess').and.callThrough();
});


test/mocks/data.js
/**
 * @author Paul.Bronshteyn
 * @comment Built by a geek loaded on caffeine ...
 */
var
  homepage = {
    "_links": {
      "curies": [
        {
          "name": "test",
          "href": "http://www.example.com/docs/{rel}",
          "templated": true
        },
        {
          "href": "invalid curie"
        }
      ],
      "self": {
        "href": "./"
      },
      "noHref": {
        "title": "No href, just title"
      },
      "noTitle": {
        "href": "no/title"
      },
      "noParams": {
        "href": "no/params/",
        "title": "Link with no params"
      },
      "noParamsNotTemplated": {
        "href": "no/params/",
        "title": "Link with no params, templated false",
        "templated": false
      },
      "noParamsTemplated": {
        "href": "no/params/",
        "title": "Link with no params, templated true",
        "templated": true
      },
      "oneParamPath": {
        "href": "one/param/{path}/",
        "title": "One param in path",
        "templated": true
      },
      "oneParamQuery": {
        "href": "one/param/{?query}",
        "title": "One param in query",
        "templated": true
      },
      "twoParamsMixed": {
        "href": "two/params/{path}/{?query}",
        "title": "Two params, one in path, one in query",
        "templated": true
      },
      "twoParamsPath": {
        "href": "two/params/{path1}/{path2}/",
        "title": "Two params in path",
        "templated": true
      },
      "twoParamsQuery": {
        "href": "two/params/{?query1,query2}",
        "title": "Two params in query",
        "templated": true
      },
      "twoParamsQueryAndExisting": {
        "href": "two/params/?test=1{&query1,query2}",
        "title": "Two params in query with hard coded param",
        "templated": true
      },
      "twoParamsQueryInvalidAndExisting": {
        "href": "two/params/?test=1{?query1,query2}",
        "title": "Two wrongly formatted params in query with hard coded param",
        "templated": true
      },
      "rootRelativeUrl": {
        "href": "/root/relative/",
        "title": "Root relative url",
        "templated": false
      },
      "pathRelativeUrl": {
        "href": "../root/relative/",
        "title": "Path relative url",
        "templated": false
      },
      "absoluteUrl": {
        "href": "http://www.example.com",
        "title": "Absolute url",
        "templated": false
      },
      "linkList": [
        {
          "href": "first/link",
          "title": "first link"
        },
        {
          "href": "second/link",
          "title": "second link"
        }
      ],
      "test:curie-test": {
        "href": "/curie/test/",
        "title": "Curie test link"
      }
    },
    "_forms": {
      "formNoAction": {
        "method": "POST",
        "fields": {}
      },
      "formNoMethod": {
        "action": "form/",
        "fields": {}
      },
      "formGetNoFields": {
        "action": "form/",
        "method": "GET"
      },
      "formGetEmptyFields": {
        "action": "form/",
        "method": "GET",
        "fields": {}
      },
      "formGetOneField": {
        "action": "form/",
        "method": "GET",
        "fields": {
          "param": {}
        }
      },
      "formGetTwoFields": {
        "action": "form/",
        "method": "GET",
        "fields": {
          "param1": {},
          "param2": {}
        }
      },
      "formGetTwoFieldsDefaultValue": {
        "action": "form/",
        "method": "GET",
        "fields": {
          "param1": {
            "default": "test"
          },
          "param2": {}
        }
      },
      "formPostNoFields": {
        "action": "form/",
        "method": "POST"
      },
      "formPostEmptyFields": {
        "action": "form/",
        "method": "POST",
        "fields": {}
      },
      "formPostOneField": {
        "action": "form/",
        "method": "POST",
        "fields": {
          "param": {}
        }
      },
      "formPostTwoFields": {
        "action": "form/",
        "method": "POST",
        "fields": {
          "param1": {},
          "param2": {}
        }
      },
      "formPostTwoFieldsDefaultValue": {
        "action": "form/",
        "method": "POST",
        "fields": {
          "param1": {
            "default": "test"
          },
          "param2": {}
        }
      },
      "formPostWithTemplatedAction": {
        "action": "form/{param}/test",
        "method": "POST",
        "fields": {
          "param1": {},
          "param2": {}
        },
        "templated": true
      }
    }
  },

  actionResponse = {
    "_links": {
      "self": {
        "href": "./"
      }
    },
    "_embedded": {
      "one": {
        "id": "636363636363",
        "title": "One Title",
        "nestedProp": {
          "name": "Prop value"
        },
        "_embedded": {
          "sub_one": null
        }
      },
      "two": [{
        "_links": {
          "self": {
            "href": "./"
          },
          "linkedEmbedded": {
            "href": "../linked"
          }
        },
        "_embedded": {
          "sub_two": {
            "_links": {
              "self": {
                "href": "../../../test/id/"
              }
            },
            "_embedded": {
              "images": [{
                "_links": {
                  "imageUrl": {
                    "href": "http://some.cdn/foo/bar.jpg"
                  }
                },
                "width": 50,
                "height": 30
              }],
              "company": [{
                "_links": {
                  "self": {
                    "href": "../../../text/company/"
                  },
                  "companyUrl": {
                    "href": "http://companyUrl"
                  }
                },
                "_embedded": {
                  "images": []
                },
                "id": "12345",
                "name": "Test Company",
                "title": "Title of company",
                "description": "Description of company"
              }]
            },
            "id": "98756",
            "title": "Sub one title"
          }
        },
        "id": "78423y08"
      }],
      "three": {
        "_links": {
          "self": "../linked"
        },
        "id": "linked123"
      }
    }
  };


test/spec/204.spec.js
/**
 * @author Paul.Bronshteyn
 * @comment Built by a geek loaded on caffeine ...
 */
describe('hyperGard', function() {
  beforeEach(function(done) {
    window.fetch.and.returnValue(this.homepageData(homepage));
    this.testHyperGard = new HyperGard(testEndpoint, testOptions);
    this.testHyperGard.fetch().then(this.onSuccess, this.onError).then(done, done);
  });

  describe('Test 204 response', function() {
    beforeEach(function(done) {
      this.link = this.data.getFirstAction('twoParamsPath', {
        path1: 'path1',
        path2: 'path2'
      });

      window.fetch.and.returnValue(this.homepageData('', 204));
      this.link.fetch().then(this.onActionSuccess, this.onActionError).then(done, done);
    });

    it('should return empty data response', function() {
      expect(this.actionData).toEqual('');
    });
  });
});


gulpfile.js
var gulp = require('gulp');
var gulpsync = require('gulp-sync')(gulp);
var jshint = require('gulp-jshint');
var rename = require("gulp-rename");
var sourcemaps = require('gulp-sourcemaps');
var lazypipe = require('lazypipe');
var karma = require('karma').Server;
var rollup = require('gulp-better-rollup');
var butternut = require('gulp-butternut');
var mergeStream = require('merge-stream');
var del = require('del');

var testFiles = [
    {
        src: './lib/extend.js',
        moduleName: 'deepExtend',
    },
    {
        src: './lib/urlparse.js',
        moduleName: 'urlparse',
    },
    {
        src: './lib/urltemplate.js',
        moduleName: 'urltemplate',
    },
    {
        src: './src/hyperGard.js',
        moduleName: 'HyperGard',
    },
];

var jshintFlow = lazypipe()
    .pipe(jshint)
    .pipe(jshint.reporter, 'jshint-stylish')
    .pipe(jshint.reporter, 'fail');

gulp.task('default', ['build-test-files', 'test-ci']);

gulp.task('test', function(done) {
    new karma({
        configFile: __dirname + '/karma.conf.js'
    }, done).start();
});

gulp.task('test-ci', function(done) {
    new karma({
        configFile: __dirname + '/karma-ci.conf.js',
        singleRun: true
    }, done).start();
});

gulp.task('jshint', function() {
    return gulp.src('src/*.js')
               .pipe(jshintFlow());
});

gulp.task('compress', function() {
    return gulp.src(['dist/*.js'])
        .pipe(butternut())
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('dist'));
})

gulp.task('clean:dist', function() {
    return del(['dist/**/*']);
})

gulp.task('bundle', function() {
    var entryPoint = './src/hyperGard.js';

    var umd = gulp.src(entryPoint)
        .pipe(sourcemaps.init())
        .pipe(rollup({
            format: 'umd',
            name: 'HyperGard',
            amd: {
                id: 'HyperGard',
            },
        }));

    var es = gulp.src(entryPoint)
        .pipe(sourcemaps.init())
        .pipe(rollup({
            format: 'es',
        }))
        .pipe(rename(function(path) {
            path.basename += ".es";
        }));

    return mergeStream([es, umd])
        .pipe(sourcemaps.write('./', { addComment: false }))
        .pipe(gulp.dest('dist'));
})

gulp.task('build-test-files', function() {
    var streams = testFiles.map(function(testFile){
        return gulp.src(testFile.src)
            .pipe(rollup({
                format: 'umd',
                name: testFile.moduleName,
                amd: {
                    id: testFile.moduleName,
                },
            }))
    })

    return mergeStream(streams)
        .pipe(gulp.dest('.test-build'));

})

gulp.task('build:all', gulpsync.sync(['jshint', 'build-test-files', 'test-ci', 'clean:dist', 'bundle', 'compress']));

gulp.task('canned', function() {
    var
        canned = require('canned'),
        http = require('http'),
        options = {
            port: '4444',
            src: './test/mocks',
            cors_headers: 'x-hypergard'
        },

        cannedOptions = {
            cors: options.cors || true,
            logger: options.logger || process.stdout,
            cors_headers: options.cors_headers || false
        },

        can = canned(options.src, cannedOptions),

        server = http.createServer(can).listen(options.port);

    console.log('Mock API server running at http://localhost:' + options.port + ', serving files from ' + options.src);
});


src/hyperGard.js
/**
 * Copyright 2018 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or   implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

import deepExtend from '../lib/extend.js';
import urltemplate from '../lib/urltemplate.js';
import urlparse from '../lib/urlparse.js';

var
  version = '6.0.0',

  defaultOptions = {
    preloadHomepage: true,
    cacheHomepage: false,
    debug: false,

    xhr: {
      headers: {
        Accept: 'application/hal+json, application/json, */*; q=0.01',
        'X-HyperGard': version
      }
    }
  },

  excludedProps = /^(_embedded|_links|_forms)$/,
  excludeBody = /^(head|get)$/i,
  concat = [].concat,
  keys = Object.keys,

  isObject = function(value) {
    return !!value && {}.toString.call(value) === '[object Object]';
  },

  xhrStatus = function(response) {
    return (response.status >= 200 && response.status < 300) ? response :
      Promise.reject(response instanceof Response ? response : new Response('', {
        status: 503,
        statusText: 'Possible CORS error'
      }));
  },

  xhrTimeout = function(timeout) {
    return new Promise(function(res, rej) {
      setTimeout(function() {
        rej({
          error: {
            code: '0011',
            msg: 'Fetch timeout',
            timeout: timeout
          }
        });
      }, timeout);
    });
  },

  load = function(url, fetchOptions) {
    var o = deepExtend({}, defaultOptions.xhr, fetchOptions);

    return Promise.race([
        xhrTimeout(fetchOptions.timeout || 60000),
        fetch(url, o)
      ]).then(xhrStatus);
  },

  urlSerialize = function(obj) {
    return Object.keys(obj).map(function(key) {
      return encodeURIComponent(key) + '=' + encodeURIComponent(obj[key]);
    }).join('&');
  },

  HyperGard = function(endpoint, initOptions) {
    var
      homepage,
      homepageLoaded = false,
      options = deepExtend({}, defaultOptions, initOptions || {}),
      linkRefs = {},

      parseCuries = function(curies) {
        var result = {};

        (curies || []).forEach(function(curie) {
          if (curie.name) {
            result[curie.name] = curie;
          }
        });

        return result;
      },

      parseEmbedded = function(res, parent) {
        var
          result = {};

        keys(res).forEach(function(key) {
          result[key] = Array.isArray(res[key]) ? res[key].map(function(item) {
            return new EmbeddedResource(item, parent);
          }) : new EmbeddedResource(res[key], parent);
        });

        return result;
      },

      sharedResourceAPI = function(obj, res, parent, base) {
        var
          props = {},
          link = ((res._links || {}).self || {}).href || '',
          actions = [],
          embedded = parseEmbedded(res._embedded || {}, obj);

        keys(res).forEach(function(key) {
          if (!excludedProps.test(key)) {
            this[key] = res[key];
          }
        }, props);

        /**
         * Get Action
         * @description Allow to follow actions (links, forms) on HAL resources and embedded resources
         * @param {String} name The name of the action to be taken
         * @param {Object} params Parameters to populate the action
         * @returns {Array} List of available actions
         */
        obj.getAction = function(name, params) {
          if (!actions[name]) {
            actions[name] = [];

            var
              linkCuries = parseCuries((res._links || {}).curies),
              formCuries = parseCuries((res._forms || {}).curies),
              curieName = String(name).split(':')[0] || '';

            concat.call([], (res._links || {})[name]).forEach(function(lnk) {
              if (lnk) {
                lnk.curie = linkCuries[curieName];
                actions[name].push(new Action(this, name, lnk, params, 'link'));
              }
            }, this);

            concat.call([], (res._forms || {})[name]).forEach(function(frm) {
              if (frm) {
                frm.curie = formCuries[curieName];
                actions[name].push(new Action(this, name, frm, params, 'form'));
              }
            }, this);
          }

          return actions[name].map(function(action) {
            return action.setParams(params);
          });
        };

        /**
         * Get base api path
         * @returns {String}
         */
        obj.getBase = function() {
          return base || parent.getBase();
        };

        /**
         * Get embedded object
         * @param {String} name Name of the embedded object
         * @returns {Object}
         */
        obj.getEmbedded = function(name) {
          return embedded[name];
        };

        /**
         * Fetch embedded
         * @description Get the object from embedded or follow a link with the same name
         * @param {String} name Name of the embedded object
         * @param {Object} [options] to provide to fetch for actions
         * @returns {Promise}
         */
        obj.fetchEmbedded = function(name, options) {
          return this.hasEmbedded(name) ? Promise.resolve({
            data: this.getEmbedded(name)
          }) : this.getFirstAction(name).fetch(options);
        };

        /**
         * Get first action from getAction list
         * @description Convenience method to get the first action instead of using getAction()[0]
         * @param {String} name The name of the action to be taken
         * @param {Object} params Parameters to populate the action
         * @returns {Object} Get Action API
         */
        obj.getFirstAction = function(name, params) {
          return this.getAction(name, params)[0] || new Action(this, name, {}, {}, 'none');
        };

        /**
         * Get parent resource object
         * @returns {Object}
         */
        obj.getParent = function() {
          return parent;
        };

        /**
         * Get property value
         * @returns {*}
         */
        obj.getProp = function(prop) {
          return props[prop];
        };

        /**
         * Get all properties as an object
         * @returns {Object}
         */
        obj.getProps = function() {
          return props;
        };

        /**
         * Does an object have a given action?
         * @description Convenience method to find out if an object has a link or form
         * @param {String} name The name of the action
         * @returns {Boolean}
         */
        obj.hasAction = function(name) {
          return obj.hasLink(name) || obj.hasForm(name);
        };

        /**
         * Does a resource contain a given embedded obj?
         * @description Convenience method to find out if a resource contains an embedded object
         * @param {String} name The name of the embedded obj
         * @returns {Boolean}
         */
        obj.hasEmbedded = function(name) {
          return !!embedded[name];
        };

        /**
         * Does an object have a given form?
         * @description Convenience method to find out if an object has a form
         * @param {String} name The name of the form
         * @returns {Boolean}
         */
        obj.hasForm = function(name) {
          return obj.listActions().forms.indexOf(name) >= 0;
        };

        /**
         * Does an object have a given link?
         * @description Convenience method to find out if an object has a link
         * @param {String} name The name of the link
         * @returns {Boolean}
         */
        obj.hasLink = function(name) {
          return obj.listActions().links.indexOf(name) >= 0;
        };

        /**
         * List of all available actions
         * @returns {Object}
         */
        obj.listActions = function() {
          var
            filter = function(item) {
              return item !== 'curies';
            };

          return {
            links: keys(res._links || {}).filter(filter),
            forms: keys(res._forms || {}).filter(filter)
          };
        };

        /**
         * List of all available embedded objects
         * @returns {Array}
         */
        obj.listEmbedded = function() {
          return keys(embedded);
        };


        if (options.debug) {
          /**
           * original HAL resource
           */
          obj.resource = res;
        }

        if (!base) {
          /**
           * Get root level resource
           */
          obj.getRoot = function() {
            var root;

            do {
              root =  this.getParent();
            } while (!(root instanceof Resource));

            return root;
          };
        }

        if (link) {
          linkRefs[link] = obj;
        }
      },

      /**
       * Action instance
       * @param {Object} parent Parent resource
       * @param {String} name Action name
       * @param {Object} action Action object
       * @param {Object} [params] Params to be used for templated action
       * @param {String} type Internal identification of action type
       * @constructor
       */
      Action = function(parent, name, action, params, type) {
        var
          api = this,
          rawUrl = (action.href || action.action || '').replace(/^#/, ''),
          method = action.method || (type === 'form' ? 'POST' : 'GET'),
          actionUrl = '',
          payLoad = '';

        /**
         * Get action name
         * @returns {String}
         */
        api.getActionName = function() {
          return name;
        };

        /**
         * Get action type
         * @returns {String}
         */
        api.getActionType = function() {
          return type;
        };

        /**
         * Get formatted url for the action
         * @returns {String}
         */
        api.getActionUrl = function() {
          return actionUrl;
        };

        /**
         * Get action method
         * @returns {String}
         */
        api.getMethod = function() {
          return method;
        };

        /**
         * Get params that need are required for the action
         * @returns {Array}
         */
        api.getParams = function() {
          return urltemplate.extractParams(rawUrl);
        };

        /**
         * Get un-formatted url for the action
         * @returns {String}
         */
        api.getRawActionUrl = function() {
          return rawUrl;
        };

        /**
         * Get action title
         * @returns {String}
         */
        api.getTitle = function() {
          return action.title || '';
        };

        /**
         * Get action template flag
         * @returns {Boolean}
         */
        api.isTemplated = function() {
          return action.templated || false;
        };

        api.setParams = function(params) {
          actionUrl = action.templated ? urltemplate.expand(rawUrl, params || {}) : rawUrl;

          if (actionUrl) {
            actionUrl = urlparse.urljoin(action.curie ? action.curie.href : parent.getBase(), actionUrl);
          }

          if (type === 'form') {
            if (action.fields && isObject(params)) {
              payLoad = {};
              keys(action.fields).forEach(function(field) {
                if (params.hasOwnProperty(field)) {
                  payLoad[field] = params[field];
                } else if (action.fields[field].hasOwnProperty('default')) {
                  payLoad[field] = action.fields[field]['default'];
                }
              });
            } else if (params) {
              payLoad = JSON.stringify(params);
            }

            if (excludeBody.test(method) && payLoad) {
              actionUrl = urlparse.urljoin(actionUrl, '?' + urlSerialize(payLoad));
              payLoad = '';
            }
          }

          return api;
        };

        if (type === 'form') {
          /**
           * Get form fields for the form action
           * @returns {Object}
           */
          api.getFields = function() {
            return action.fields || {};
          };

          /**
           * Get payload to be submitted for the form action
           * @returns {Object}
           */
          api.getPayload = function() {
            return payLoad;
          };
        }

        if (action.curie) {
          /**
           * Get curie documentation url
           * @returns {String}
           */
          api.getDocUrl = function() {
            return urltemplate.expand(action.curie.href, {
              rel: name.split(':')[1]
            });
          };
        }

        api.setParams(params);
      },

      /**
       * HAL Embedded resource
       * @param {Resource} res HAL resource
       * @param {Resource} parent Parent HAL resource
       * @constructor
       */
      EmbeddedResource = function(res, parent) {
        return sharedResourceAPI(this, res || {}, parent);
      },

      /**
       * HAL Resource
       * @param {String} base Base url for the resource
       * @param {Object} res Resource data
       * @constructor
       */
      Resource = function(base, res) {
        return sharedResourceAPI(this, res, this, base);
      };

    Action.prototype.fetch = function(fetchOptions) {
      fetchOptions || (fetchOptions = {});

      var
        url = fetchOptions.url || this.getActionUrl(),
        rawUrl = this.getRawActionUrl(),
        name = this.getActionName(),
        payLoad = this.getPayload ? this.getPayload() : '',
        o = deepExtend({
          method: this.getMethod(),
          action: name,
        }, options.xhr, fetchOptions),

        onSuccess = function(response) {
          if (response.status === 204) {
            return Promise.resolve({
              action: name,
              data: '',
              xhr: response
            });
          }

          /**
           * If flag to returnRawData is false & the object is a valid
           * use Resource constructor
           * @param {} data
           */
          function formatData(data) {
            if (!o.returnRawData && isObject(data)) {
              return new Resource(url, data);
            }

            return data;
          }

          return response.json().then(function(data) {
            return {
              action: name,
              data: formatData(data),
              xhr: response
            };
          }, function() {
            return {
              action: name,
              data: response.text(),
              xhr: response
            };
          });
        },

        onError = function(response) {
          return Promise.reject(response instanceof Response ? {
            error: {
              action: name,
              code: '0021',
              msg: 'Failed to retrieve action'
            },
            xhr: response
          } : response);
        };

      if (!excludeBody.test(o.method) && isObject(payLoad)) {
        o.headers['Content-Type'] = 'application/x-www-form-urlencoded';
        o.body = urlSerialize(payLoad);
      }

      if (!url) {
        return Promise.reject({
          error: {
            action: name,
            code: '0020',
            msg: 'Url is not provided for this action'
          }
        });
      } else if (o.method === 'GET' && !fetchOptions.force && linkRefs.hasOwnProperty(rawUrl)) {
        return Promise.resolve({
          action: name,
          data: linkRefs[rawUrl]
        });
      }

      return load(url, o).then(onSuccess, onError);
    };

    /**
     * Fetch
     * @returns {Promise}
     */
    this.fetch = function() {
      var o = deepExtend({
        method: 'GET',
        action: 'homepage',
      }, options.xhr);
      var
        onSuccess = function(response) {
          if (!options.cacheHomepage) {
            homepageLoaded = false;
          }

          return response.json().then(function(data) {
            return isObject(data) ? {
              data: new Resource(endpoint, data),
              xhr: response
            } : Promise.reject();
          })['catch'](function() {
            return Promise.reject({
              error: {
                code: '0002',
                msg: 'Could not parse homepage'
              },
              xhr: response
            });
          });
        },

        onError = function(response) {
          homepageLoaded = false;

          return Promise.reject(response instanceof Response ? {
            error: {
              code: '0001',
              msg: 'Failed to retrieve homepage'
            },
            xhr: response
          } : response);
        };

      if (!endpoint) {
        return Promise.reject({
          error: {
            code: '0000',
            msg: 'API endpoint was not provided'
          }
        });
      }

      if (!homepageLoaded) {
        homepageLoaded = true;
        homepage = load(endpoint, o).then(onSuccess, onError);
      }

      return homepage;
    };

    /**
     * Set global options
     * @param {Object} o Options object
     * @return {Object} HyperGard
     */
    this.setOptions = function(o) {
      deepExtend(options, o || {});
      return this;
    };

    /**
     * Get global options
     * @returns {Object}
     */
    this.getOptions = function() {
      return options;
    };

    /**
     * Parse json into resource
     * @param {Object} json Data in json format to be parsed as resource
     * @returns {Resource}
     */
    this.parse = function(json) {
      return new Resource(endpoint, json || {});
    };

    /**
     * Fetch homepage on initialization
     */
    if (options.preloadHomepage) {
      this.fetch();
    }
  };

HyperGard.prototype.version = version;

/**
 * Will wrap any fetch performed in supplied middleware
 * This will allow custom logging headers to be set without
 * using before/after fetch events
 * @param {Function} middleware Function to wrap fetches in
 */
function applyMiddleware(middleware) {
  load = (function(stack) {
    return function(url, fetchOptions) {
      return middleware(url, fetchOptions, stack);
    };
  })(load);
}

/**
 * Will apply an array of middleware functions around the load method
 * @param {Array} middlewareStack Array of functions to be wrapped around every load
 */
HyperGard.prototype.applyMiddlewareStack = function(middlewareStack) {
  var applicationOrder;
  if (middlewareStack && Array.isArray(middlewareStack) && middlewareStack.length) {
    // Apply in reverse order, for first entry is applied as inner most wrapper
    applicationOrder = middlewareStack.reverse();
    applicationOrder.forEach(applyMiddleware);
  }
};


export default HyperGard;

